HW0/P6/P6.txt

```diff
@@ -1 +1,51 @@
 # Your answers here
+PART A.
+    Example output:
+        Hi Job 0
+        Hi Job 1
+        Hi Job 2
+        Hi Job 3
+        Bye Job 0
+        Bye Job 1
+        Bye Job 3
+        Bye Job 2
+        Hi Job 4
+        Hi Job 5
+        Hi Job 6
+        Hi Job 7
+        Bye Job 4
+        Bye Job 6
+        Bye Job 7
+        Bye Job 5
+        Hi Job 8
+        Hi Job 9
+        Bye Job 9
+        Bye Job 8
+        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+    
+    The jobs do not always complete in the order they were started (even though
+    they nominally sleep for the same amount of time). This means we can't rely
+    on processes running in parallel completing in any particular order. If two
+    processes modify the same resource, for example, and one relies on the
+    output of the other, unexpected behavior may result (e.g., if a bank is
+    processing many transactions in parallel and a user deposits their paycheck
+    and then pays a bill, the deposit may not have completed before the bill
+    payment is deducted from the user's account, resulting in overdraft fees).
+
+PART B.
+    The speedup of parallel relative to serial execution increases as the wait
+    time increases, finally plateauing at roughly the theoretical maximum of
+    4x speedup for wait times on the order of 10^-2 through 10^0 seconds.
```

Comment:
Yeah, in practice the serial running time isn't completely fixed, so if on one iteration the serial process was slowed down (other concurrent processes seems like a good guess for why) and the parallel processes weren't, they could be more than 4x faster. If we ran many more iterations with the same wait time and averaged them, we could get a smoother curve that wouldn't go above 4.0, but this is a pretty small sample size so it's noisy.